### PR TITLE
Addison Cole - Hotel - C10 - Edges

### PR DESCRIPTION
**Pull request by addisoncole**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Certain design choices I am proud of: the use of the Helper Method class which holds mostly organizing/sorting methods, binary search methods and tests for errors. They are general house cleaning methods for data. I am excited for my incorporation of 2 binary searches that do different things, which I felt was 1) a good design choice as data became larger this would be an efficient way of finding data, and 2) a good reuse of single use methods which required no design changes to the overall structure just plug and play in the places and in the kinds of cases I needed. I also chose to use Room Class and make rooms and object, though I do feel they could've been incorporated into the booking manager class as structs maybe or just as an array of hashes. However, I like its current state for ease of access of data, however the trade off in its current iteration is a little weird in my midn which is that each room stores their own reservations which seems counterintuitive for the booking manager and I feel it may be more appropriate for the bookign manager to keep track of all of them. I did have an overall instance variable in booking manager of ALL reservations, but it was not as conducive to sorting and finding all of the reservations so I removed it in the end because it was not being used. it would be easy to add in as future feature if needed.
Describe a concept that you gained more clarity on as you worked on this assignment. | I feel like writing unit tests completely from scratch and using TDD was really helpful in not over designing. I also really appreciated the use of namespacing and feel like I understand class methods vs instance methods in better way.
Describe a _nominal test_ that you wrote for this assignment. | Being able to book a room on day when no reservations were previously reserved
Describe an _edge case test_ that you wrote for this assignment. | Testing a for being able to book a reservation in a room on the same day a previous reservation was checking out. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I actually found I wrote less pseudocode this project than I have any in the last two but I did write an overwhelming amount of my unit tests before writing my code. The peseudocode i wrote was often around edge cases and the more complex methods. I did find that some of the more complex methods were harder to write code on and sometimes I found that it seemed like it was harder when they had more dependencies on other methods.

### CODE REVIEW

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | x
**Wave 1** |
List rooms | 
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | x
Test coverage | x
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | x
Test coverage | x
**Wave 3** |
Create a block of rooms | x
Check if a block has rooms | x
Reserve a room from a block | x
Test coverage | x
**Fundamentals** | 
Names variables, classes and modules appropriately | Making a few comments on how to improve, but otherwise fantastic
Understanding of variable scope - local vs instance | Making a few comments on literally a few cases, but otherwise fantastic
Can create complex logical structures utilizing variables | YES!
Appropriately uses methods to break down tasks into smaller simpler tasks | YES!
Understands the differences between class and instance methods | x
Appropriately uses iterators and enumerables | YES!
Appropriately writes and utilizes classes | x
Appropriately utilizes modules as a mechanism to encapsulate functionality | YES!
**Wrap Up** |
There is a refactors.txt file | x
The file provides a roadmap to future changes | x, I would agree with maybe revisiting what gets into the HelperMethod class and what goes into BookingManager
**Additional Feedback** | 

Addison!!!!! I can feel from reading your code how you had fun with this project and how you got to explore some really fun things.
I absolutely love the `HelperMethod` class you made. I think that your BookingManager had a really solid foundation and leveraged the HelperMethod class really well and it paid off. I also think you set up Room, Reservation, and BlockRoom well as objects that held information, with very little behavior in general.

I think that your tests are detailed, deep, extensive, and accurate, too!

I'm leaving a bunch of comments, but the themes overall are:
1. Some thoughts about some naming things
1. Things that I think are really cool!
1. Questions about potential ways to move around code
1. Questions to get you to think about maybe minor refactorings to make existing functionality a different code style

Overall, your Hotel project is really great and a joy-- I can tell that a lot of the details were really thoughtful. You obviously had priorities to use Ruby to make your code DRY, to be efficient, to be FUN, to be you, and to be elegant.

Lastly, your comprehension question answers imply that you were pretty intentional about your process to approaching this project. I encourage you to reflect on what worked well and to keep doing it when it makes sense!

By the way, if you haven't already shared with Shruti your implementation of binary search here, you absolutely should!!

File: lib/booking_manager.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+require_relative 'block_room'
+require_relative 'helper_method'
+
+module Hotel
+  class Booking_Manager
+
+    attr_reader :rooms, :hotel_reservations, :block_reservations
+
+    ROOM_NUMBERS = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
+    ROOM_COST = 200.0
```
Comment (tildeee) [inline lib/booking_manager.rb:13]: YAY CONSTANTS!

File: lib/booking_manager.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    def list_reservations(date)
+      reservations = []
+      @rooms.each do |room|
+        index = Hotel::Helper_Method.binary_search_reservations_return_index_if_found(room.reservations,date)
+        if index == nil
+          next
+        elsif
+          reservations << room.reservations[index]
+        end
```
Comment (tildeee) [inline lib/booking_manager.rb:41]: Hm. This `if ... elsif` feels weird, does it work to turn it into `if index != nil`?

File: lib/booking_manager.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        blocked_rooms << reservation
+      end
+      @block_reservations << blocked_rooms
+      Hotel::Helper_Method.sort_block_reservations_by_name(@block_reservations)
+    end
+
+    def reserve_room_in_block(input)
+      customer_name = input[:name]
+      available_rooms_in_block = check_block_status(input[:block_name])
```
Comment (tildeee) [inline lib/booking_manager.rb:80]: Does the method `check_block_status` have an accurate name if the result of it is the available rooms?

File: lib/booking_manager.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    end
+
+### BOOKING UTILITY METHODS ###
+    def total_cost_of_stay(reservation)
+      cost = (reservation.nights_of_stay.length * ROOM_COST)
+      if reservation.class == Hotel::Block_Room
+        return (cost - (cost * reservation.block_discount))
+      else
+        return cost
+      end
```
Comment (tildeee) [inline lib/booking_manager.rb:104]: is there a way we can move the logic of calculating `cost` into the `reservation`? And if so, is there a way we can use inheritance to cover the case if reservation is a `Block_Room` too?

File: lib/booking_manager.rb [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    end
+
+    def search_all_rooms_availability(check_in_date, check_out_date)
+      possible_nights_of_stay = Hotel::Helper_Method.generate_nights(check_in_date, check_out_date)
+      vacant_rooms = []
+      @rooms.each do |room|
+        if Hotel::Helper_Method.binary_search_list_of_reservations_for_vacancy(room.reservations, possible_nights_of_stay) == false
+          vacant_rooms << room.room_number
+        end
+      end
```
Comment (tildeee) [inline lib/booking_manager.rb:114]: I think a future refactoring may be to change this `each` (lines 110-114) to a `.filter` call

File: lib/booking_manager.rb [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        return raise ArgumentError, 'No Rooms Available in date range'
+      end
+      return vacant_rooms
+    end
+
+    def get_rooms
+      return @rooms
+    end
+
+    private
```
Comment (tildeee) [inline lib/booking_manager.rb:125]: Nice! We didn't teach this

File: lib/helper_method.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
+module Hotel
+  class Helper_Method
```
Comment (tildeee) [inline lib/helper_method.rb:2]: In software, one existing design pattern that this class and its responsibility might align with/fit with is the idea of a Utility Class. I might call this `HotelUtility` instead of `Helper_Method`.

Also, naming convention in Ruby for classes is TitleCase/PascalCase/CapitalCamelCase, so a better name than `Helper_Method` is `HelperMethod`!

File: lib/helper_method.rb [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+### SORTING METHODS ###
+    def self.sort_reservations_by_date(list_of_reservations)
+      list_of_reservations.sort! { |a,b| a.check_in_date <=> b.check_in_date }
+    end
```
Comment (tildeee) [inline lib/helper_method.rb:7]: I LOVE EVERYTHING IN THIS CLASS

File: lib/helper_method.rb [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+    def self.sort_block_reservations_by_name(list_of_block_reservations)
+      # SORTS AN ARRAY OF ARRAYS BY NAME
+      list_of_block_reservations.sort! { |a,b| a.first.block_name <=> b.first.block_name }
+    end
+
+    def self.connect_reservation_to_room_and_sort(list_of_rooms, room_number, reservation)
+      connected_room_number = Hotel::Helper_Method.find_room_number(list_of_rooms, room_number)
```
Comment (tildeee) [inline lib/helper_method.rb:15]: By the way, Ruby will always check for local names first, so there's no need to namespace fully... this should still work even if it's just `connected_room_number = find_room_number(list_of_rooms, room_number)`

File: lib/helper_method.rb [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        if !(array_of_possible_dates & array_of_reservations[mid].nights_of_stay).empty?
+          return true
+        elsif array_of_reservations[mid].check_in_date > array_of_possible_dates.first
+          max = mid - 1
+        elsif array_of_reservations[mid].check_in_date < array_of_possible_dates.first
+          min = mid + 1
+        end
+      end
+
+      if !(array_of_possible_dates & array_of_reservations[0].nights_of_stay).empty?
```
Comment (tildeee) [inline lib/helper_method.rb:44]: Just want to confirm: Do you want the binary `&` operator and not the AND `&&`?

File: lib/helper_method.rb [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    end
+
+    def self.binary_search_reservations_return_index_if_found(array_of_reservations, date)
+      # GIVEN A DATE, IT RETURNS INDEX OF A RESERVATION ON THAT DATE, NIL IF NO RESERVATIONS
+      min = 0
+      max = array_of_reservations.length
+
+      if array_of_reservations.first == nil
+        return nil
+      end
```
Comment (tildeee) [inline lib/helper_method.rb:58]: would love this `if` block as a one liner as a refactoring ;)

File: lib/reservation.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      check_out = input[:check_out_date]
+      Hotel::Helper_Method.check_valid_date_range(check_in, check_out)
+
+      nights_stay = Hotel::Helper_Method.generate_nights(check_in, check_out)
+
+      @name = input[:name]
+      @room_number = input[:room_number]
+      @check_in_date = input[:check_in_date]
+      @check_out_date = input[:check_out_date]
+      @nights_of_stay = nights_stay
```
Comment (tildeee) [inline lib/reservation.rb:17]: Maybe go ahead and inline this so it's `@nights_of_stay = Hotel::Helper_Method.generate_nights(check_in, check_out)`?

File: spec/booking_manager_spec.rb [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+      @input = { name: "Dr. Frankenstein",
+        room_number: 6,
+        check_in_date: @date1,
+        check_out_date: @date2
+      }
+    end
+
+    it 'adds a reservation to the array of all reservations for a room' do
+      @hotel.reserve_room(@input)
+      expect(@hotel.rooms[5].reservations[0].name).must_equal "Dr. Frankenstein"
```
Comment (tildeee) [inline spec/booking_manager_spec.rb:54]: As much as I love Dr. Frankenstein gettin' a reservation at the ole Hotel Ada, is there a requirement in the project to have a name associated?

File: spec/booking_manager_spec.rb [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+    end
+
+    it 'checks against room availbility and returns an Argument error if the date range is filled.' do
+      @all_rooms_available.each do |i|
+        @input2 = { name: "Señor Dracula",
```
Comment (tildeee) [inline spec/booking_manager_spec.rb:59]: I think `@input2` can be a local variable `input2` in this case! Also THESE NAMES

File: spec/booking_manager_spec.rb [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+  describe 'list_reservations method' do
+    before do
+      @hotel = Hotel::Booking_Manager.new
+      @date = Date.new(2018, 6, 8)
+      @input1 = { name: "Dr. Frankenstein",
+        room_number: 6,
+        check_in_date: Date.new(2018, 6, 6),
+        check_out_date: Date.new(2018, 6, 9)
+      }
+      @input2 = { name: "Mx. Mummy",
```
Comment (tildeee) [inline spec/booking_manager_spec.rb:79]: I'M SCREAMING, I LOVE A MX. VERY VERY VVVVV MUCH :) :) :) :)))))))) :))))))))))) ok yes ty for bringing life into code with good test data

File: spec/booking_manager_spec.rb [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+      @hotel.reserve_room_in_block(@input3)
+      @check2 = @hotel.check_block_status(@input2)
+      expect(@check2.length).must_equal 2
+    end
+
+    it 'returns an error if all rooms in block booked' do
+      @hotel.reserve_room_in_block(@input3)
+      @hotel.reserve_room_in_block(@input3)
+      @hotel.reserve_room_in_block(@input3)
+      expect{ @hotel.check_block_status(@input2) }.must_raise StandardError
```
Comment (tildeee) [inline spec/booking_manager_spec.rb:297]: I feel like I would want this error to be raised if I tried to call `@hotel.reserve_room_in_block(@input3)` again instead of `check_block_status`?

I think this raises questions for me about what the responsibility of the method `check_block_status` is
